sql: check use of virtual columns in indexes

RaduBerinde opened the pull request:
This commit adds checks that disallow virtual columns being used as
primary key columns or stored columns in indexes.

Release note: None

Informs #57608.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57998)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Eventually both of these things seem handy. 

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57998#-:-MOhoaCA9WoMaXEttcPE:b-z074j8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR! Agreed, we can extend as needed (we just need to clarify the desired semantics).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57998#-:-MOhscelAcaCNt7J5osl:b-qvg3xh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: check use of virtual columns in indexes

This commit adds checks that disallow virtual columns being used as
primary key columns or stored columns in indexes.

Release note: None

RaduBerinde commented:
bors r+
